cudax/include/cuda/experimental/__execution/stream/domain.cuh

```diff
@@ -0,0 +1,73 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __CUDAX__EXECUTION_STREAM_DOMAIN
+#define __CUDAX__EXECUTION_STREAM_DOMAIN
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/experimental/__execution/domain.cuh>
+
+#include <cuda/experimental/__execution/prologue.cuh>
+
+namespace cuda::experimental::execution
+{
+//////////////////////////////////////////////////////////////////////////////////////////
+// stream evaluation function dispatched on algorithm tag
+#if defined(__CUDA_ARCH__) && _CCCL_OS(WINDOWS)
+// "A __device__ variable template cannot have a const qualified type on Windows"
+template <class _Tag>
+__device__ auto __stream_apply = default_domain::__apply(_Tag());
+#else
+template <class _Tag>
+_CCCL_GLOBAL_CONSTANT auto __stream_apply = default_domain::__apply(_Tag());
+#endif
+
+//////////////////////////////////////////////////////////////////////////////////////////
+// stream domain
+struct stream_domain : default_domain
+{
+  template <class _Tag>
+  _CCCL_TRIVIAL_HOST_API static constexpr auto __apply(_Tag) noexcept
+  {
+    return __stream_apply<_Tag>;
+  }
+
+  _CUDAX_SEMI_PRIVATE :
+  template <class _Tag>
+  struct __apply_t;
+
+  struct __visitor;
+};
+
+/////////////////////////////////////////////////////////////////////////////////
+// __visitor: walk a sender expression and turn it into something that can be
+// enqueued on a stream
+struct stream_domain::__visitor
+{
+  template <class _Context, class _Tag, class _Params, class... _Child>
+  _CCCL_TRIVIAL_HOST_API auto operator()(_Context& __context, _Tag, _Params&& __params, _Child&&... __child)
+  {
+    return __stream_apply<_Tag>(__context, static_cast<_Params&&>(__params), static_cast<_Child&&>(__child)...);
+  }
+};
```

Comment:
Question: Why is this defined outside of the class?

Reply:
no particular reason. shall i move it?

Reply:
If you have other stuff that warrants a push then yes, otherwise no

Reply:
The reason I am asking is that I like to follow Kate Gregory's advice about how to structure code.

This stands out to me so I am wondering why it is different then other definitions.

Reply:
> I like to follow Kate Gregory's advice about how to structure code.

@miscco where can i find her advice?

Reply:
Its from her CppCon Talks, I believe its https://www.youtube.com/watch?v=n0Ak6xtVXno